### PR TITLE
Improved the way requirements are handled, ensuring that pip gets all the information it needs.

### PR DESCRIPTION
**Pull request by pelson**
Also renames the pypi name from Iris to scitools-iris.

### CODE REVIEW

**Comment by pelson** [comment]
Note: There would need to be some follow up work to ensure the developer/installation instructions are accurate.

**Comment by pelson** [comment]
Note that this requires a non-old setuptools and/or pip. I was getting:

```
(iris_dev) pelson@~/dev/scitools/iris> python setup.py install
error in scitools-iris setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/version requirement specifiers.
```

```
(iris_dev) pelson@~/dev/scitools/iris> vim requirements/full.txt 
(iris_dev) pelson@~/dev/scitools/iris> conda update pip setuptools
Fetching package metadata .....
Solving package specifications: .

Package plan for installation in environment /Users/pelson/miniconda/envs/iris_dev:

The following packages will be UPDATED:

    pip:        8.0.1-py27_0 defaults --> 9.0.1-py27_0  conda-forge
    setuptools: 19.4-py27_0  defaults --> 36.6.0-py27_1 conda-forge
```

**I can then install with setup.py**

File: .travis.yml [hunk 1]
```diff
@@ -55,16 +55,34 @@ install:
...
+          python requirements/gen_conda_requirements.py --py2 --groups test full > ${CONDA_REQS_FILE};
       fi
+      cat ${CONDA_REQS_FILE};
+      conda install --quiet --file ${CONDA_REQS_FILE};
     fi
 
+
+  - >
+    if [[ "${TEST_TARGET}" == 'doctest' ]]; then
+        conda install --quiet --file requirements/docs.txt
```
Comment (bjlittle) [inline .travis.yml]: @pelson `requirements/docs.txt` doesn't seem to exist ...

File: requirements/setup.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Dependencies necessary to run setup.py of iris
+# ----------------------------------------------
+
+setuptools
+numpy
```
Comment (bjlittle) [inline requirements/setup.txt]: Do we need `numpy` ?

File: setup.py [hunk 3]
```diff
@@ -68,6 +67,22 @@ def temporary_path(directory):
 with temporary_path('lib/iris/tests/runner'):
     from _runner import TestRunner  # noqa:
 
+SETUP_DIR = os.path.dirname(__file__)
+
+def requirements(name):
```
Comment (bjlittle) [inline setup.py]: rename to `pip_requirements` ?

File: requirements/gen_conda_requirements.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    else:
+        conda_reqs_lines.append('# Python 3 conda configuration')
+
+    for req_name in requirement_names:
+        fname = os.path.join(REQS_DIR, '{}.txt'.format(req_name))
+        if not os.path.exists(fname):
+            raise RuntimeError('Unable to find the requirements file for {} '
+                               'in {}'.format(req_name, fname))
+        conda_reqs_lines.extend(read_conda_reqs(fname, options))
+        conda_reqs_lines.append('')
```
Comment (bjlittle) [inline requirements/gen_conda_requirements.py]: Why do you need this?

File: requirements/full.txt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+
+# Iris extensions (i.e. key tools that depend on Iris)
+# ----------------------------------------------------
+
+# Note: pip can handle the circularity of these extensions, but conda will
+# struggle. To install these extensions, ensure iris[core] has been installed
+# first.
+
+iris_grib;python_version<"3"  #conda:
```
Comment (bjlittle) [inline requirements/full.txt]: Would it be worth while (since we're inventing syntax) making this explicitly `#conda: ignore` ?

File: setup.py [hunk 6]
```diff
@@ -214,16 +228,27 @@ def extract_version():
...
+
     zip_safe=False,
-   )
+
+    setup_requires=requirements('setup'),
+    install_requires=requirements('core'),
+    tests_require=['{}[test]'.format(pypi_name)],
+    extras_require = {
+                      'test': requirements('test'),
+                      'all': requirements('full'),
```
Comment (bjlittle) [inline setup.py]: `all` or `full` ?

File: .travis.yml (hunk 1 above)
Comment (pelson) [reply .travis.yml]: Done.

File: requirements/setup.txt (hunk 2 above)
Comment (pelson) [reply requirements/setup.txt]: No. But we do need: pyke and six.

Comment (pelson) [reply requirements/setup.txt]: 🤕  - pyke not pip installable.

Comment (pelson) [reply requirements/setup.txt]: I've raised on the pyke mailing-list.

File: requirements/gen_conda_requirements.py (hunk 4 above)
Comment (pelson) [reply requirements/gen_conda_requirements.py]: Simply some whitespace.

File: requirements/full.txt (hunk 5 above)
Comment (pelson) [reply requirements/full.txt]: If you like... 😉 . What if there is a package on conda called ignore though... how would we install that!?! 😄

File: setup.py (hunk 6 above)
Comment (pelson) [reply setup.py]: Well, I guess I prefer all...

**Comment by pelson** [comment]
New commit added.

File: requirements/extensions.txt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# Iris extensions (i.e. key tools that depend on Iris)
+# ----------------------------------------------------
+
+# Note: pip can handle the circularity of these extensions, but conda will
+# struggle. To install these extensions, ensure iris[core] has been installed
+# first.
+
+iris_grib;python_version<"3"  #conda:
```
Comment (bjlittle) [inline requirements/extensions.txt]: @pelson the `extensions.txt` doesn't seem to be used anywhere. Intentional?

File: setup.py [hunk 8]
```diff
@@ -214,16 +228,27 @@ def extract_version():
...
+
     zip_safe=False,
-   )
+
+    setup_requires=pip_requirements('setup'),
+    install_requires=pip_requirements('core'),
+    tests_require=['{}[test]'.format(pypi_name)],
+    extras_require = {
+                      'test': pip_requirements('test'),
+                      'all': pip_requirements('full'),
```
Comment (bjlittle) [inline setup.py]: This should now be `pip_requirements('all')` ... right?

File: .travis.yml [hunk 9]
```diff
@@ -55,15 +55,25 @@ install:
...
+    CONDA_REQS_FILE=conda-requirements.txt
+    python requirements/gen_conda_requirements.py ${CONDA_REQ_GROUPS} ${CONDA_REQ_FLAGS} > ${CONDA_REQS_FILE}
+    cat ${CONDA_REQS_FILE}
+    conda install --quiet --file ${CONDA_REQS_FILE}
+
+  - >
+    if [[ "${TEST_TARGET}" == 'doctest' ]]; then
+        conda install --quiet --file requirements/docs.txt
+    fi
+
```
Comment (bjlittle) [inline .travis.yml]: @pelson Travis doesn't like your refactor ... :cry:

File: requirements/extensions.txt (hunk 7 above)
Comment (pelson) [reply requirements/extensions.txt]: Ja. It could happily live as an extra (e.g. ``pip install scitools-iris[extensions]``)...

**Comment by pelson** [comment]
Phew. I'm a few iterations down, but have now ensured that the test-data was being installed properly (it wasn't in my previous passing commit). Ideally we'd have a check to ensure there are no skipped tests, but I didn't get around to implementing that.

Sorry for the added noise to ``.travis.yml``. The way we do requirements is highly bound to the way we do continuous integration, and it was inevitable that the two would collide in this PR...

Should now be good to go.

File: .travis.yml [hunk 10]
```diff
@@ -24,80 +24,97 @@ env:
...
+    echo "doc_dir = $(pwd)/docs/iris" >> $SITE_CFG;
+    echo "[System]" >> $SITE_CFG;
+    echo "udunits2_path = $PREFIX/lib/libudunits2.so" >> $SITE_CFG;
 
   - python setup.py --quiet install
 
+  # JUST FOR NOW : Install latest version of iris-grib.
+  # TODO : remove when iris doesn't do an integration test requiring iris-grib.
+  - if [[ "$TEST_MINIMAL" != true ]]; then
+        conda install python-eccodes
```
Comment (bjlittle) [inline .travis.yml]: @pelson You missed a trailing semi-colon, which is upsetting travis ...

**Comment by bjlittle** [comment]
@pelson We're now having systemic grib failures ...

**Comment by pelson** [comment]
> curiously, this default test is failing test_license_headers (iris.tests.test_coding_standards.TestLicenseHeaders) ... SKIP: Iris installation did not look like a git repo

Doesn't that always happen in our existing tests too? If not, I'll dig a little...

**Comment by pelson** [comment]
Confirmed. On master we are currently:

 * skipping iris-grib tests on default (an oversight that I'm going to fix in this PR)
 * skipping dot tests on "default" (they are not being done anywhere by the looks of it...)
 * skipping the license header tests on "default" (they are done in coding)

**Comment by pelson** [comment]
```
======================================================================

FAIL: test_projections_and_annotations (example_tests.test_projections_and_annotations.TestProjectionsAndAnnotations)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/SciTools/iris/docs/iris/example_tests/test_projections_and_annotations.py", line 37, in test_projections_and_annotations

    projections_and_annotations.main()

  File "/home/travis/build/SciTools/iris/docs/iris/example_code/General/projections_and_annotations.py", line 110, in main

    make_plot('Equidistant Cylindrical', ccrs.PlateCarree())

  File "/home/travis/build/SciTools/iris/docs/iris/example_code/General/projections_and_annotations.py", line 105, in make_plot

    iplt.show()

  File "/home/travis/build/SciTools/iris/docs/iris/example_tests/extest_util.py", line 71, in replacement_show

    test_case.check_graphic()

  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/scitools_iris-2.0a0-py3.6.egg/iris/tests/__init__.py", line 842, in check_graphic

    raise AssertionError(emsg.format(msg))

AssertionError: Image comparison failed: Bad phash 5fa17298a1580c27336eb3d08d9e4c3aed61a68d931c93d273ce7939ece4ad03 with hamming distance [8, 8, 6] for test example_tests.test_projections_and_annotations.TestProjectionsAndAnnotations.test_projections_and_annotations.0.
```

😢 - I'm going to dig now, but I have no idea what the issue is here... 😭

**Comment by pelson** [comment]
Looks like it is entirely unrelated, and would have happened on master too (confirming that now). I *think* it is related to GEOS being pinned on conda-forge: https://github.com/conda-forge/cartopy-feedstock/pull/41

I'm going to add the new image and we can move forwards (it is fine):

diff

**Comment by pelson** [comment]
Confirmed. This is a failing test on master (as a result, I think, of the GEOS update).

**Comment by bjlittle** [comment]
@pelson https://github.com/SciTools/iris/pull/2912 is merged, so you can rebase ...

**Comment by pelson** [comment]
Do you actually want me to rebase? Are you happy with the implication that you will need to re-review the changes?

**Comment by pelson** [comment]
P.S. Instead of rebasing, I've given the power-cycle a shot (can't quite remember what that does in terms of CI testing)...

**Comment by pelson** [comment]
Looking promising. I've cherry-picked the stickler config into the v2.0.x branch, and pushed to the repo. I've also merged v2.0.x back into master. Will do the same once this is merged.

**Comment by bjlittle** [comment]
@pelson LGTM! Awesome work 👍

**Comment by pelson** [comment]
Thanks @bjlittle!

**Pull request merged**